// @observability-hub/observability/dist/index.js
"use strict";
// ==============================================
// @observability-hub/observability
// Complete Observability Solution for Microservices
// ==============================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.createObservabilitySetup = exports.createReadinessCheckHandler = exports.createHealthCheckHandler = exports.createRedisService = exports.createRedisClient = exports.RedisClient = exports.createDerivedConfig = exports.createObservabilityConfigSchema = exports.ObservabilityLogger = exports.initTracer = void 0;
const logger_1 = require("./logger/logger");
const redis_1 = require("./redis");
const health_1 = require("./health");
const config_1 = require("./config");
// ===== TRACING EXPORTS =====
var tracing_1 = require("./tracing");
Object.defineProperty(exports, "initTracer", { enumerable: true, get: function () { return tracing_1.initTracer; } });
// ===== LOGGER EXPORTS =====
var logger_2 = require("./logger/logger");
Object.defineProperty(exports, "ObservabilityLogger", { enumerable: true, get: function () { return logger_2.ObservabilityLogger; } });
// ===== CONFIGURATION EXPORTS =====
var config_2 = require("./config");
Object.defineProperty(exports, "createObservabilityConfigSchema", { enumerable: true, get: function () { return config_2.createObservabilityConfigSchema; } });
Object.defineProperty(exports, "createDerivedConfig", { enumerable: true, get: function () { return config_2.createDerivedConfig; } });
// ===== REDIS EXPORTS =====
var redis_2 = require("./redis");
Object.defineProperty(exports, "RedisClient", { enumerable: true, get: function () { return redis_2.RedisClient; } });
Object.defineProperty(exports, "createRedisClient", { enumerable: true, get: function () { return redis_2.createRedisClient; } });
Object.defineProperty(exports, "createRedisService", { enumerable: true, get: function () { return redis_2.createRedisService; } });
// ===== HEALTH CHECK EXPORTS =====
var health_2 = require("./health");
Object.defineProperty(exports, "createHealthCheckHandler", { enumerable: true, get: function () { return health_2.createHealthCheckHandler; } });
Object.defineProperty(exports, "createReadinessCheckHandler", { enumerable: true, get: function () { return health_2.createReadinessCheckHandler; } });
// ===== FACTORY FUNCTION =====
const createObservabilitySetup = (serviceName, config = {}) => {
    const logger = new logger_1.ObservabilityLogger({
        serviceName,
        serviceVersion: config.serviceVersion || '1.0.0',
        environment: config.environment || 'development',
        ...config.logger,
    });
    return {
        logger,
        redis: config.redis ? (0, redis_1.createRedisService)(config.redis, serviceName, logger) : null,
        healthCheck: (0, health_1.createHealthCheckHandler)(serviceName, config.serviceVersion || '1.0.0', {}),
        config: (0, config_1.createObservabilityConfigSchema)(serviceName, 8080, 9090),
    };
};
exports.createObservabilitySetup = createObservabilitySetup;
// ===== DEFAULT EXPORTS =====
exports.default = {
    ObservabilityLogger: logger_1.ObservabilityLogger,
    createObservabilityConfigSchema: config_1.createObservabilityConfigSchema,
    createRedisClient: redis_1.createRedisClient,
    createHealthCheckHandler: health_1.createHealthCheckHandler,
    createObservabilitySetup: exports.createObservabilitySetup
};
